level-3

naruls opened the pull request:
first commit

naruls committed:
Add files via upload

first commit

naruls committed:
Update README.md

Update readme

naruls committed:
Update README.md

Update README

naruls committed:
Update README.md

Update README

naruls committed:
Update README.md

irinamozes commented:
Здравствуйте, Кирилл.

В Вашем чек-листе написано:

### Работа отклоняется от проверки

- ....

- Фронтенд-часть приложения не выложена на сервер.

**Вы должны предоставить ссылку на выложенную на сервер работающую фронтенд часть приложения. Напишите эту ссылку, пожалуйста, в файле README.md.**

Должна проверяться работоспособность версии проекта, выложенной на сервер.

**Напоминаю:** не забудьте, что надо на каждую проверку присылать архив с последней версией кода, находящейся в Вашем репозитории и в облаке.
Укажите, в файле README.md архива проекта **ссылку на пулл реквест этапа реакта, ссылку на фронтенд-часть проекта и ссылку на макет.**

Текущая итерация засчитываться не будет без потери количества итераций.

naruls committed:
Update README.md

naruls committed:
исправления

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит страницу фильмов, или перейдёт на неё с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска (с расстановкой актуального состояния лайков), который был перед перезагрузкой или уходом на другую страницу. При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать. Сейчас пользователь не видит правильного результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы - при переходе со страницы «Сохранённые фильмы» видит карточки, которые были найдены на этой странице. После перезагрузки сайта видит пустой экран, а не результат последнего поиска.**
    
     **Комментарий: Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью независимы друг от друга.**
     
     **Комментарий: при логауте пользователя нужно всю информацию вышедшего пользователя в локальном хранилище и также обязательно в стейт-переменных из useState уничтожать. Иначе, информация  пользователя, сделавшего логаут, может стать известна вошедшему следом другому пользователю, не говоря о том, что и сам проект в некоторых случаях будет работать неправильно. Сейчас, если следующий пользователь авторизуется на сайте без предварительной перезагрузки сайта после логаута предыдущего пользователя, ему при входе на сайт на странице фильмов отображается поиск предыдущего пользователя. Видимо, эта информация не уничтожается в стейт-переменных из какого-то useState при логауте пользователя. Кейс необходимо протестировать, можно на пользователях review1@gmail.com и review2@gmail.com, пароль 12345678. При этом нужно проверить разные варианты входа пользователей на сайт: когда пользователь авторизуется на сайте без предварительной перезагрузки сайта после логаута предыдущего пользователя, и когда он авторизуется на сайте после предварительной перезагрузки сайта после логаута предыдущего пользователя.**
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: При вводе в поле email адреса почты без точки перед именем домена верхнего уровня (в правой части адреса) можно выполнить запрос на сервер, что противоречит текущему требованию чек-листа, так как в этом случае адрес будет невалидноным. Нужно, чтобы валидация имени, почты и пароля и на серверной и на клиентской частях приложения, была одинаковой (по одним и тем же регулярным выражениям), и при невалидных данных запрос на сервер вообще не возможно было отправить.**
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    **Комментарий: Если запрос прошёл успешно, то происходит редирект на страницу авторизации, а нужен автоматический вход на `/movies` как авторизованного пользователя, без необходимости посещения страницы авторизации.**
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    **Комментарий: Пользователю не отображается уведомление об успешном запросе к серверу при сохранении профиля.**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
	**Комментарий: При редактировании пользователь может ввести текущие данные профиля и сохранить их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если  данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте.**
	
	**Напоминание: после каждого успешного сохранения данных профиля на сервере, нужно эти данные вносить в контекст юзера.**
	
	**Замечание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
    **Комментарий: многие карточки нельзя сохранить из-за того, что сервер приложения присылает в ответе ошибку 400 с сообщением о невалидности данных (см. снимок). Но, на сервере приложения не должно быть проверки валидности атрибутов карточки (кроме, возможно, проверки правильности типа атрибута), поскольку эти данные там не вводятся, как и не вводятся они на клиенте. Проверка данных осуществляется на сервере beatfilm по установленным им критериям, и предполагается, что он отдаёт валидные данные (если вообще их отдаёт). Поэтому на клиенте должна быть возможность сохранения всех карточек.**
![2023-05-15_18-49](https://github.com/naruls/movies-explorer-frontend/assets/19558891/acda3ec3-c89d-45ca-8e9f-c0412aac7439)
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: При попытке авторизованного пользователя войти на любую страницу по её URL-адресу в адресной строке, происходит вход пользователя на страницу «Фильмы», а не на эти страницы, хотя редирект не по набранному адресу должен происходить только при попытке авторизованного пользователя выйти на страницы авторизации и регистрации.**
  - [ ] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  **Комментарий: После успешного вызова обработчика `onSignOut` происходит редирект на страницу авторизации, а не на `/`.**
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные числовые данные нужно заносить в константы с понятными названиями и помещать в файл констант, чтобы в коде не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [ ] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  **Комментарий: На страницах присутствуют сообщения, которые относятся к другим страницам, или уже не нужны (см. снимок). На страницах должны присутствовать только актуальные в текущий момент надписи, сообщения нужно убирать, когда они становятся неактуальными.**
![Снимок экрана в 2023-05-15 20-00-34](https://github.com/naruls/movies-explorer-frontend/assets/19558891/5e8bce98-5090-45a4-aa74-a4d7041868af)
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: Если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а  не пустой экран и не результат какого-либо поиска.**
## Количество баллов: 53.
4 балла снято за замечание из комментария в разделе "Доп. Комментарии".
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет. Это обязательное требование.**

naruls committed:
Разделён функционал поиска фильмов, реализован показ найденных фильмов после перезагрузки

Разделён функционал поиска фильмов, реализован показ найденных фильмов после перезагрузки

naruls committed:
При выходе необходимые state переменные "чистятся"

При выходе необходимые state переменные "чистятся"

naruls committed:
Добавлена валидация email

Добавлена валидация email

naruls committed:
Добавлена валидация email

Добавлена валидация email

naruls committed:
Добавлен автоматический вход после регистрации

Добавлен автоматический вход после регистрации

naruls committed:
Добавлено уведомление об изменении профиля, добавленна валидация email в update profile 

Добавлено уведомление об изменении профиля, добавленна валидация email в update profile

naruls committed:
Исправлена ошибка, из-за которой пользователь мог отправить на сервер данные профиля без изменений

Исправлена ошибка, из-за которой пользователь мог отправить на сервер данные профиля без изменений

naruls committed:
Исправлена ошибка, из-за которой авторизированный пользователь всегда перенаправлялся на /movies

naruls committed:
Исправлено перенаправление после выхода из аккаунта 

Исправлено перенаправление после выхода из аккаунта

naruls committed:
Исправлена ошибка с сохранением записи об ошибки после перехода на другой путь (signin, signup)

Исправлена ошибка с сохранением записи об ошибки после перехода на другой путь (signin, signup)

naruls committed:
Добавлены константы

naruls committed:
Добавлено закрытие по нажатию оверлея

naruls committed:
Исправлена ошибка, из-за которой при перезагрузке страницы вместо имени и email были undefined

Исправлена ошибка, из-за которой при перезагрузке страницы вместо имени и email были undefined

naruls committed:
Общие исправления

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.    
    **Комментарий: Сохранять данные  поиска на странице «Сохранённые фильмы» в localstorag не нужно. Сейчас эти данные зпносятся в localstorage в те же ключи name и value, что и данные со страницы фильмов, заменяя их значения, что ведёт к неправильному повторному показу результата поиска на странице фильмов. Нужно, чтобы при входе на страницу «Сохранённые фильмы» пользователь видел все свои избранные карточки без задания параметров поиска (с параметрами поиска в дефолтном состоянии).**
    
     **Комментарий: Как было сказано при предыдущей проверке, функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью независимы друг от друга.**
     
      **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит страницу фильмов, или перейдёт на неё с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска (с расстановкой актуального состояния лайков), который был перед перезагрузкой или уходом на другую страницу. При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.**
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    **Комментарий: Если запрос прошёл успешно, то происходит редирект на страницу авторизации, а нужен автоматический вход на `/movies` как авторизованного пользователя, без необходимости посещения страницы авторизации.**
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    **Комментарий: на странице «Сохранённые фильмы» должен выводиться результат поиска, или выводиться все избранные карточки при переходе на неё, без использования кнопки «Ещё», а полностью всем массивом за один раз. Сейчас кнопка «Ещё» присутствует на экране страницы «Сохранённые фильмы» (см. снимок).**
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо, так как открывается возможность авторизоваться ещё раз, не выйдя из предыдущего аккаунта.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: Если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а  не пустой экран и не результат какого-либо поиска.**
## Количество баллов: 59.
4 балла снято за замечание из комментария в разделе "Доп. Комментарии".
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет. Это обязательное требование.**

naruls committed:
При переходе на сохраненные фильмы отображаются сохраненные фильмы

При переходе на сохраненные фильмы отображаются сохраненные фильмы

naruls committed:
Зарегистрированный пользователь теперь не может вернуться на страницу регистрации/входа

Зарегистрированный пользователь теперь не может вернуться на страницу регистрации/входа

naruls committed:
Разделен функционал поиска movies и saved-movies

Разделен функционал поиска movies и saved-movies

naruls committed:
Разделен функционал поиска movies и saved-movies

Разделен функционал поиска movies и saved-movies

naruls committed:
Разделен функционал поиска movies и saved-movies

Разделен функционал поиска movies и saved-movies

naruls committed:
Исправлена ошибка, из-за которой при переходе из saved-movies в movies после поиска в saved-movies ломалось приложение

Исправлена ошибка, из-за которой при переходе из saved-movies в movies после поиска в saved-movies ломалось приложение

naruls committed:
Исправлена ошибка с изменением цвета лого при наведении курсором на область рядом с лого

Исправлена ошибка с изменением цвета лого при наведении курсором на область рядом с лого

naruls committed:
Исправлена ошибка лого

Исправлена ошибка лого

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.    
    **Комментарий: после перезагрузки страницы фильмов или перехода на неё с другой страницы пропадает строка поиска в форме поиска (см. снимок).**
![Снимок экрана в 2023-05-23 23-55-48](https://github.com/naruls/movies-explorer-frontend/assets/19558891/47643537-fedb-41ef-a116-7fd022990707)
      **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит страницу фильмов, или перейдёт на неё с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска (с расстановкой актуального состояния лайков), который был перед перезагрузкой или уходом на другую страницу. При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.**
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: Если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их. Сейчас все фильмы показываются только после перезагрузки страницы «Сохранённые фильмы», в других случаях при переходе на эту страницу показывается результат последнего поиска, если он был перед уходом.**
## Количество баллов: 63.
2 балла снято за замечание из комментария в разделе "Доп. Комментарии".
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет. Это обязательное требование.**

naruls committed:
Добавлено выставление значения checkbox и строки поиска после перезагрузки/перехода на страницу movies

Добавлено выставление значения checkbox и строки поиска после перезагрузки/перехода на страницу movies

naruls committed:
Исправлена ошибка, из-за которой не высвечивалось сообщение "не найдено" при поиски с чекбоксом 

Исправлена ошибка, из-за которой не высвечивалось сообщение "не найдено" при поиски с чекбоксом

naruls committed:
При переходе на страницу с сохраненными фильмами отображаются все сохраненные фильмы (без перезагрузки)

При переходе на страницу с сохраненными фильмами отображаются все сохраненные фильмы (без перезагрузки)

naruls committed:
Исправлено отображение лого при разрешении меньше 1280 px

Исправлено отображение лого при разрешении меньше 1280 px

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.    
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Количество баллов: 100.
## У Вас отличный дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Руководство Typescript
http://typescript-lang.ru/docs/

TypeScript в деталях.
https://typescript-definitive-guide.ru/book/chapters

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**
